k2/csrc/utils.cu

```diff
@@ -312,17 +352,169 @@ void RowIdsToRowSplits(ContextPtr &c, int32_t num_elems, const int32_t *row_ids,
-          num_rows, threads_per_elem, row_ids, num_elems, row_splits));
+          threads_per_elem= RoundUpToNearestPowerOfTwo(avg_rows_per_elem),
+          tot_threads = num_elems * threads_per_elem;
+      int32_t dim_block = 256;
+      int32_t dim_grid = NumBlocks(tot_threads, dim_block);
+      RowIdsToRowSplitsKernel<<<dim_block, dim_grid, 0, c->GetCudaStream()>>>(
+          num_elems, threads_per_elem,
+          row_ids, num_rows,
+          row_splits);
+      K2_CHECK_CUDA_ERROR(cudaGetLastError());
     }
   }
 }
 
+/*
+  See declaration of RowSplitsToRowIds() in utils.h.  This is implementation notes.
+
+    Suppose the range we need to fill with a
+    particular number (say, x) is from 1010 to 10000 inclusive (binary) The
+    first kernel writes x to positions 1010, 1100, 10000; the significance of
+    that sequence is we keep adding the smallest number we can add to get
+    another zero at the end of the binary representation, until we exceed the
+    range we're supposed to fill.  The second kernel: for a given index into x
+    that is must fill (say, 1111), it asks "is the index currently here already
+    the right one?", which it can test using the function is_valid_index()
+    below; if it's not already corret, it searches in a sequence of positions:
+    1110, 1100, 1000, 0000, like our sequence above but going downwards, again
+    getting more zeros at the end of the binary representation, until it finds
+    the correct value in the array at the searched position; then it copies the
+    discovered value the original position requested (here, 1111).
+
+
+    First kernel pseudocode: for each index 'i' into 't', it does:
+      for (int32_t n=0, j = t[i]; j < t[i+1]; n++) {
+         x[j] = i;
+         if (j & (1<<n))  j += (1 << n);
+      }
+    Second kernel pseudocode: for each element of x, it searches for the right index.  Suppose we're
+    given num_indexes == length(n) == length(t) - 1.  Define is_valid_index as follows:
+       // returns true if j is the value that we should be putting at position 'i' in x:
+       // that is, if t[j] <= i < t[j+1].
+       bool is_valid_index(i, j) {
+          return (j >= 0 && j < num_indexes && t[j] <= i && i < t[j+1]);
+       }
+       // We suppose we are given i (the position into x that we're responsible for
+       // setting:
+       orig_i = i;
+       for (int32_t n=0; !is_valid_index(i, x[i]); n++) {
+         if (i & (1<<n))  i -= (1 << n);
+       }
+       x[orig_i] = x[i];
+*/
+void RowSplitsToRowIds(ContextPtr &c, int32_t num_rows, const int32_t *row_splits,
+                       int32_t num_elems, int32_t *row_ids) {
+  DeviceType d = c->GetDeviceType();
+  if (d == kCpu) {
+    int32_t cur_row_start = row_splits[0];
+    K2_CHECK_EQ(cur_row_start, 0);
+    for (int32_t row = 0; row < num_rows; row++) {
+      int32_t next_row_start = row_splits[row+1];
+      for (; cur_row_start < next_row_start; ++cur_row_start)
```

Comment:
just curious,.. why is the color here red, anyone knows the reason?

Reply:
I think this is caused by line 359. GitHub cannot handle `<<< >>>` correctly.